remove dockerfile-build and install deps directly

nilsnolde opened the pull request:
ref https://github.com/valhalla/valhalla/issues/3685#issuecomment-1199818890

simplifies to only one Dockerfile and removes the `docker` directory. Dockerfile-build was merged with Dockerfile-run and circleci jobs install their own deps now.

nilsnolde commented on .circleci/config.yml:
```diff
@@ -13,42 +13,33 @@ commands:
   build-docker-images:
-    docker:
-      - image: docker:18.06.3-ce-git
+    machine:
+      image: ubuntu-2004:2022.07.1
```
we can use the circleci provided image so it doesn't download: circleci.com/developer/machine/image/ubuntu-2204#how-this-image-works

nilsnolde replied:
switch to `current` ubuntu tag

nilsnolde commented on .circleci/config.yml:
```diff
@@ -13,42 +13,33 @@ commands:
       - checkout
       - run:
-          name: Build Valhalla Docker Images
+          name: Build and push Valhalla Docker image
```
do it all in one go

nilsnolde commented on .circleci/config.yml:
```diff
@@ -13,43 +13,35 @@ commands:
       - checkout
+      - install_linux_dependencies
       - run: |
+            python3 -m pip install requests
```
confusingly, circleci doesn't use the same python version as installed by our script.. makes no sense and is very annoying

nilsnolde commented on scripts/needs_ci_run:
```diff
@@ -128,6 +128,9 @@ for changed_file in changed_files:
         if fnmatch.fnmatch(changed_file, ignore_rule):
             ignored_files.append(changed_file)
 
+print(f"changed files: {changed_files}")
+print(f"ignored files: {ignored_files}")
```
I actually didn't know this was still active.. I haven't ever seen it skip anything, I wanna see what it thinks are changed & ignored files

nilsnolde commented:
stupid circleci has serious python3 issues, I'll look at it later

ImreSamu commented on scripts/install-linux-deps.sh:
```diff
@@ -56,4 +56,4 @@ env DEBIAN_FRONTEND=noninteractive apt-get install --yes --quiet \
 
 # for boost
-pip install conan
+python3 -m pip install conan
```
imho: more secure with `--upgrade` 
* `python3 -m pip install --upgrade conan`

so we can remove the same conan upgrade lines from the new "Dockerfile"  ( added by me )

ImreSamu commented on Dockerfile:
```diff
@@ -1,11 +1,15 @@
+
+# get the code into the right place, install deps and prepare to build it
 WORKDIR /usr/local/src/valhalla
 ADD . /usr/local/src/valhalla
+RUN bash ./scripts/install-linux-deps.sh
```
imho: for the optimal docker layer caching 
the order of these two lines should be reversed.

proposal:
```
RUN bash ./scripts/install-linux-deps.sh
ADD . /usr/local/src/valhalla
```
and please add a [`.dockerignore` file](https://docs.docker.com/engine/reference/builder/#dockerignore-file)   :smile: 
   * ignoring a `./build` directory - makes the local docker image testing and building faster.

nilsnolde replied:
> the order of these two lines should be reversed.

jep that's true (just gotta add the script before using it).

I'll add a .dockerignore as well. I never build this image myself, but it hurts if you do

ellenhp commented:
I haven't thoroughly reviewed but I like this a lot in principle. I'm doing some work on the docker image as well to get a native arm64 image and ended up doing many of these same things independently. It'll be a pain to rebase my changes but I think it'll be worth it :)

nilsnolde commented on .circleci/config.yml:
```diff
@@ -12,44 +12,35 @@ commands:
-              docker push valhalla/valhalla:build-latest
-              docker push valhalla/valhalla:run-latest
-            elif [[ ! -z "${CIRCLE_TAG}" ]]; then
+              BUILD_TAG="${CIRCLE_TAG:-latest}"
```
now we'd only build docker images on master and releases, but IMO that's fine, the other CI builds run on very similar systems

nilsnolde commented on .dockerignore:
```diff
@@ -0,0 +1,69 @@
+# same as .gitignore
```
I had this on my laptop since a while apparently, it's just a copy of .gitignore. could be overhauled I guess, but meeh

nilsnolde commented on Dockerfile:
```diff
@@ -1,19 +1,22 @@
 
-# get the code into the right place and prepare to build it
+# set paths
+ENV PATH /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:$PATH
+ENV LD_LIBRARY_PATH /usr/local/lib:/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/lib32:/usr/lib32
```
never tried without those extra paths. they don't seem needed anymore on 20.04?

nilsnolde commented on Dockerfile:
```diff
@@ -1,19 +1,22 @@
+
+# install deps
 WORKDIR /usr/local/src/valhalla
+COPY ./scripts/install-linux-deps.sh /usr/local/src/valhalla/scripts
+RUN bash ./scripts/install-linux-deps.sh
```
first copy and execute the deps script for better layer caching as @ImreSamu pointed out. this is effectively doing a similar job as having a separate build docker image, esp if (hopefully soon) remote layer caching is standard by default

nilsnolde commented on scripts/needs_ci_run:
```diff
@@ -1,12 +1,11 @@
-#!/usr/bin/env python3
+#!/usr/bin/python3
```
hmpf, took me longer than it should've to understand what was happening.. in circleci we have to use `sudo` to install the dependencies (also python3-requests). it uses `/usr/bin/python3` for that. this script is being run without `sudo` and references `/opt/circleci/.pyenv/versions/3.10.5/bin/python3` inside circleci's virtual env and apparently they didn't bother to include `/usr/lib/python3.10/site-packages` in the PYTHONPATH, so importing `requests` fails.. 

since it only runs in ci I think it's fair to use this line instead

nilsnolde commented on .circleci/config.yml:
```diff
@@ -12,44 +12,36 @@ commands:
     steps:
       - run: brew install protobuf cmake ccache libtool libspatialite pkg-config luajit curl wget czmq lz4 spatialite-tools unzip
-      - run: pip3 install requests shapely conan
+      - run: /usr/bin/python3 -m pip install --user requests
+      - run: pip3 install shapely conan
```
`requests` is used by the `needs_ci_run` script which runs with `/usr/bin/python3`.

the other two libraries are used by cmake & valhalla and are run with `python3` which is apparently different from `/usr/bin/python3` so we need to install with that one as before

nilsnolde commented:
what a pain python can be.. this should be ready now.. to summarize:
- this removes the docker `build-*` images and instead we use a normal ubuntu ci image and install the dependencies on each job, which is only takes 40 secs per job
- we don't build any docker image anymore on PRs, only on `master` pushes and releases. as mentioned above, it's fine IMO bcs the circleci and docker base image are ubuntu 20.04. building the build image took around 2 minutes, so it's a net zero on ci time pretty much
- plain `Dockerfile`, removed `docker` dir
- moved `install_linux_deps.sh` script to `scripts`

nilsnolde commented on .circleci/config.yml:
```diff
@@ -12,44 +12,36 @@ commands:
   build-docker-images:
-    docker:
-      - image: docker:18.06.3-ce-git
+    machine:
+      image: ubuntu-2004:current
```
this should have a current docker version installed: https://circleci.com/developer/machine/image/ubuntu-2004#how-this-image-works. also `build-essentials` and others, so installing deps will take less time

nilsnolde commented on Dockerfile:
```diff
@@ -1,19 +1,22 @@
 RUN git submodule sync && git submodule update --init --recursive
 RUN rm -rf build && mkdir build
 
-# upgrade Conan:
-# https://github.com/valhalla/valhalla/issues/3685#issuecomment-1198604174
-RUN pip install --upgrade conan
-
```
on the other hand @ImreSamu : if we remove this line, we wouldn't actually prevent issue #3685 for docker builds. as long as the installation script doesn't change, locally cached layers will be used. I think it's best to keep it

ImreSamu replied:
> locally cached layers will be used.

the ["old" documentation](https://github.com/valhalla/valhalla/blob/3ecf1c7d87854e9317113c1f67504f95a7b4ddc8/docker/README.md#to-buildpublish-images-manually
)  contains the `--no-cache` suggestion, probably we should also add the `--pull`
*  `docker build --pull --no-cache ...`
   *  `--no-cache`  "Do not use cache when building the image"
   *  `--pull` : "Always attempt to pull a newer version of the image"

probably the circle-CI `build-docker-images` job should contain these 2 extra safety parameters.
* now: `docker build --tag valhalla/valhalla:run-${BUILD_TAG} .`
* proposed: `docker build  --pull --no-cache  --tag valhalla/valhalla:run-${BUILD_TAG} .`

nilsnolde replied:
I'm not so concerned about circleci builds actually, more about user builds bcs we don't cache docker in ci, but other machines do. though we could pull the newest base image, true, I'll add that.

kevinkreiser reviewed (approved):
too bad ci can't conditionally run the docker stuff, that it has to still run and pretend like it did something
